ambry-replication/src/main/java/com/github/ambry/replication/CloudToStoreReplicationManager.java

```diff
@@ -240,8 +240,13 @@ private void removeCloudReplica(String partitionName) {
   /**
    * Randomly select a {@link DataNodeId} from list of {@code vcrNodes}.
    * @return randomly selected {@link DataNodeId} object.
+   * @throws ReplicationException if there are no vcr nodes.
    */
-  private DataNodeId getCloudDataNode() {
+  private DataNodeId getCloudDataNode() throws ReplicationException {
+    if (vcrNodes.get().size() == 0) {
+      logger.error("No vcr node found to replicate partition from cloud.");
+      throw new ReplicationException("No vcr node found to replicate partition from cloud.");
+    }
     return vcrNodes.get().toArray(new CloudDataNode[0])[Utils.getRandomShort(new Random()) % vcrNodes.get().size()];
```

Comment:
I realize this isn't new code, but it seems inefficient for a getter method to have to copy a collection to an array every time it needs to get an element from it.  It makes me wonder if the concurrent set is really the optimal data structure for this.

Reply:
It's fine to address last comment in a future PR.

Reply:
You are right. It makes sense to make it a list. We modify it only in the `CloudToStoreReplicationManager::handleChangeInVcrNodes()` method and access it only in the `CloudToStoreReplicationManager::getCloudDataNode()`. And it makes sense to modify the atomic reference directly.  Have made the change.